sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala

```diff
@@ -283,83 +336,72 @@ case class GpuProjectAstExec(
-    case Nil => Array.emptyBooleanArray
-    case curTier :: Nil =>
-      // For the last tier, fill with all true (skip) values, because the output
-      // of the last tier does not include any input columns.
-      Array.fill(curTier.attrs.size)(true)
-    case curTier :: tail =>
-      val curAttrs = curTier.attrs
-      val nextAttrs = tail.head.attrs
-      // This is equivalent to:
-      // skipList = curAttrs.map(a => if (nextAttrs.contains(a)) false else true)
-      // but this should be faster
-      val skipList = new Array[Boolean](curAttrs.size)
-      var curIdx = 0
-      var nextIdx = 0
-      while (curIdx < curAttrs.size) {
-        if (nextAttrs(nextIdx) == curAttrs(curIdx)) {
-          skipList(curIdx) = false
-          nextIdx += 1
-        } else {
-          skipList(curIdx) = true
-        }
-        curIdx += 1
-      }
-      skipList
+ case class GpuTieredProject(exprTiers: Seq[Seq[GpuExpression]]) extends Arm {
+
+  /**
+   * Is everything deterministic. This can help with reliability in the common case.
+   */
+  private lazy val areAllDeterministic = !exprTiers.exists { tier =>
+    tier.exists { expr =>
+      !expr.deterministic
+    }
   }
 
-  def tieredProject(batch: ColumnarBatch): ColumnarBatch = {
-    @tailrec
-    def recurse(boundExprs: Seq[Seq[GpuExpression]], attrTiers: Seq[AttributeSeq],
-        cb: ColumnarBatch, isFirst: Boolean): ColumnarBatch = boundExprs match {
-      case Nil =>
-        if (isFirst) {
-          // if there are no bound expressions, return an empty ColumnarBatch.
-          new ColumnarBatch(Array.empty, cb.numRows())
-        } else {
-          cb
-        }
-      case exprSet :: tail =>
-        val projectCb = withResource(new NvtxRange("project tier", NvtxColor.ORANGE)) { _ =>
-          closeOnExcept(GpuProjectExec.project(cb, exprSet)) { projectResult =>
-            projectResult
-          }
+  def projectAndCloseWithRetrySingleBatch(sb: SpillableColumnarBatch,
+      spillCallback: SpillCallback): ColumnarBatch = {
+    if (areAllDeterministic) {
+      // If all of the expressions are deterministic we can just run everything and retry it
+      // at the top level. If some things are non-deterministic we need to split them up and
+      // do the processing in a way that makes it so retries are more likely to succeed.
+      RmmRapidsRetryIterator.withRetryNoSplit(sb) { sb =>
+        withResource(sb.getColumnarBatch()) { cb =>
+          project(cb)
         }
-        val nextCb = if (tail.isEmpty) {
-          projectCb
-        } else {
-          val columnSkips = getColumnSkips(attrTiers)
-          withResource(projectCb) { newCols =>
-            withResource(GpuColumnVector.dropColumns(cb, columnSkips)) { remainingCb =>
-              GpuColumnVector.combineColumns(remainingCb, newCols)
-            }
+      }
+    } else {
+      @tailrec
+      def recurse(boundExprs: Seq[Seq[GpuExpression]],
```

Comment:
(nit) maybe stylistic, but I'd prefer if this recurse inner function were moved up to the top of this function.